File: auth-service/src/main/java/lib/quick/authservice/global/security/handler/CustomAccessDeniedHandler.java
package lib.quick.authservice.global.security.handler;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lib.quick.authservice.global.security.dto.ExceptionResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;
import org.springframework.web.ErrorResponse;

import java.io.IOException;

@Slf4j
@Component
@RequiredArgsConstructor
public class CustomAccessDeniedHandler implements AccessDeniedHandler {
    private final ObjectMapper objectMapper;

    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException {
        HttpStatus forbidden = HttpStatus.FORBIDDEN;

        String responseString = objectMapper.writeValueAsString(
            new ExceptionResponse(forbidden.value(), "해당 엔드포인트에 대한 권한이 없습니다.")
        );

        log.error("{}", responseString);

        response.setCharacterEncoding("UTF-8");
        response.setStatus(forbidden.value());
        response.setContentType(MediaType.APPLICATION_JSON_VALUE);
        response.getWriter().write(responseString);
    }
}

File: auth-service-spring/src/main/kotlin/lib/quick/authservice/global/security/handler/CustomAccessDeniedHandler.kt
package lib.quick.authservice.global.security.handler;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lib.quick.authservice.global.security.dto.ExceptionResponse;
import lib.quick.authservice.global.util.getLogger
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component;

@Component
class CustomAccessDeniedHandler(
    private val objectMapper: ObjectMapper
): AccessDeniedHandler {
    override fun handle(request: HttpServletRequest, response: HttpServletResponse, accessDeniedException: AccessDeniedException) {
        val forbidden = HttpStatus.FORBIDDEN;

        val responseString = objectMapper.writeValueAsString(
            ExceptionResponse(forbidden.value(), "해당 엔드포인트에 대한 권한이 없습니다.")
        );

        getLogger().error("CustomAccessDeniedHandler: {}", responseString);

        response.characterEncoding = "UTF-8";
        response.status = forbidden.value();
        response.contentType = MediaType.APPLICATION_JSON_VALUE;
        response.writer.write(responseString);
    }
}
